src/util/futurepool.rs

```diff
@@ -0,0 +1,259 @@
+}
+
+impl<T: Context> Clone for ContextDelegators<T> {
+    fn clone(&self) -> Self {
+        ContextDelegators::<T> {
+            delegators: sync::Arc::clone(&self.delegators),
+        }
+    }
+}
+
+impl<T: Context> ContextDelegators<T> {
+    fn new(delegators: HashMap<thread::ThreadId, ContextDelegator<T>>) -> Self {
+        ContextDelegators::<T> {
+            delegators: sync::Arc::new(delegators),
+        }
+    }
+
+    fn get_current_thread_delegator(&self) -> &ContextDelegator<T> {
+        let thread_id = thread::current().id();
+        if let Some(delegator) = self.delegators.get(&thread_id) {
+            delegator
+        } else {
+            unreachable!();
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A future thread pool that supports `on_tick` for each thread.
+pub struct FuturePool<T: Context> {
+    pool: cpupool::CpuPool,
+    context_delegators: ContextDelegators<T>,
+}
+
+impl<T: Context> Clone for FuturePool<T> {
+    fn clone(&self) -> FuturePool<T> {
+        FuturePool::<T> {
+            pool: self.pool.clone(),
+            context_delegators: self.context_delegators.clone(),
+        }
+    }
+}
+
+impl<T: Context> util::AssertSend for FuturePool<T> {}
+impl<T: Context> util::AssertSync for FuturePool<T> {}
+
+impl<T: Context> FuturePool<T> {
+    pub fn new<F>(
+        pool_size: usize,
+        stack_size: usize,
+        name_prefix: &str,
+        tick_interval: time::Duration,
+        context_factory: F,
+    ) -> FuturePool<T>
+    where
+        F: Send + 'static + Fn(thread::ThreadId) -> T,
+    {
+        let (tx, rx) = mpsc::sync_channel(pool_size);
+        let pool = cpupool::Builder::new()
+            .pool_size(pool_size)
+            .stack_size(stack_size)
+            .name_prefix(name_prefix)
+            .after_start(move || {
+                // We only need to know each thread's id and we can build context later
+                // by invoking `context_factory` in a !Sync way.
+                let thread_id = thread::current().id();
+                tx.send(thread_id).unwrap();
+            })
+            .create();
+        let contexts = (0..pool_size)
+            .map(|_| {
+                let thread_id = rx.recv().unwrap();
+                let context = context_factory(thread_id);
+                let context_delegator = ContextDelegator::new(context, tick_interval);
+                (thread_id, context_delegator)
+            })
+            .collect();
+        FuturePool {
+            pool,
+            context_delegators: ContextDelegators::new(contexts),
+        }
+    }
+
+    pub fn spawn<F>(&self, f: F) -> cpupool::CpuFuture<F::Item, F::Error>
+    where
+        F: Future + Send + 'static,
+        F::Item: Send + 'static,
+        F::Error: Send + 'static,
+    {
+        // TODO: Support busy check
+        let delegators = self.context_delegators.clone();
+        let f = f.then(move |r| {
```

Comment:
how to use the context in the future?

Reply:
This requires changing `future` to a `future_factory`, like [this](https://github.com/breeswish/tikv/compare/wenxuan/readpool...breeswish:wenxuan/readpool-integration#diff-a2828e4f71806f3e4623d048057803d4R586). Because there are no usage in this PR, for simplicity and reducing confusing purpose, I kept `future` instead of `future_factory` in this PR.